ramp4-pcar4/ramp4-pcar4
# Move map caption functionality to MapCaption API

## Closes #662 

## Changes in this PR
- [FEAT] Created new MapCaption API nugget: `$iApi.geo.map.caption`
- [FEAT] Previous map caption functionality (updating attribution and scale bar) has been moved from the `MapAPI` to this new `MapCaptionAPI`
- [FEAT] Implemented "Fancy Formatting Suite" custom point formatter design from discussion #575 
- [FEAT] Created default point formatters:
    - DMS
    - DDM
    - DD
    - Web_Mercator
    - Canada_Atlas_Lambert 
    - UTM
    - Basemap
- [FIX] Fixed warnings thrown by formatter due to invalid points returned by `projectGeometry` (#666 👻)
    - Note that this is fixed by manually wrapping values in the formatter
    - The `ProjectionAPI` still needs to be fixed so that it does not return invalid coordinates 

## Further work/comments/questions
- **Further Work:**
    - Support French number formatting: #668
        - For some point formats (e.g. Mercator/Lambert), the values can be more than 3 digits
            - Values like `1,000` should be formatted as `1 000` (though currently the values are not displayed with commas)
        - Some point formats (e.g. DD or DDM) display decimal values
            - Values like `1,000.21` should be formatted as `1 000,21`
        - Suggesting to set up [Vue `i18n` number localization](https://kazupon.github.io/vue-i18n/guide/number.html)  so that it can be used throughout the app
    - Specify default formatter in the config: #670 

## [Demo](http://ramp4-app.azureedge.net/demo/users/sharvenp/662/host/index.html)
### Steps to test:
1. Load Demo
2. Mouse around the map - the default point format should be DMS
3. Change the point format by running this line:
```js
window.rInstance.geo.map.caption.setPointFormatter(<FORMATTER_ID>)
// FORMATTER_ID can be one of:
//     "LAT_LONG_DMS"
//     "LAT_LONG_DD"
//     "LAT_LONG_DDM"
//     "WEB_MERCATOR"
//     "CANADA_ATLAS_LAMBERT"
//     "UTM"
//     "BASEMAP"
``` 
4. Mouse around the map and check if the new format is working
5. Change the point format to a newly defined function:
```js
window.rInstance.geo.map.caption.setPointFormatter((p) => Promise.resolve(`X: ${p.x}, Y: ${p.y}`))
```
6. Mouse around the map and check if the new function is working


---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/ramp4-pcar4/ramp4-pcar4/667)
